cpp/src/arrow/filesystem/filesystem.h

```diff
@@ -519,6 +574,82 @@ Result<std::shared_ptr<FileSystem>> FileSystemFromUriOrPath(
 
 /// @}
 
+/// \defgroup filesystem-factory-registration Helpers for FileSystem registration
+///
+/// @{
+
+/// \brief Register a Filesystem factory
+///
+/// Support for custom Uri schemes can be added by registering a factory
+/// for the corresponding FileSystem.
+///
+/// \param[in] scheme a Uri scheme which the factory will handle.
+///            If a factory has already been registered for a scheme,
+///            the new factory will be ignored.
```

Comment:
Since this is mostly called from start up code where there isn't a Status/Result stack to unwind, I thought it'd be acceptable to simply log or ignore name conflicts. One alternative would be to store any errors and raise them in future calls to FileSystemFromUri. Is that preferable?

Reply:
At some point it may be called in other contexts as well. For example a Python module that provides such a factory? There's no reason not to be flexible here IMHO.

Reply:
Alright, how about:
- `RegisterFileSystemFactory(defer_errors=false)` raises an error directly
- `RegisterFileSystemFactory(defer_errors=true)` never raises, instead any errors will be raised by `fs::Initialize()`

Reply:
Sounds good, should we make `defer_errors=false` the default?

Reply:
Actually I think I've got a better idea. I can associate deferred errors with the specific scheme in the registry. Then FileSystemFromUri against uris with that scheme will raise the deferred error but uris of any other scheme will function normally

Reply:
This is orthogonal with wanting an eager error return, though.

Reply:
I imagined that part was non-negotiable :)